### PR TITLE
#428 calendar view of activities

### PR DESCRIPTION
**Pull request by camianderson**
## Proposed changes

Display the calendar in the calendar route of the Frontend.
Inside of the calendar is displayed all the activities for the current program.

## Screenshots
Screen Shot 2022-11-22 at 4 16 56 PM
Screen Shot 2022-11-22 at 4 17 07 PM

## Next steps
- Create the logic for when the user click the activity in the calendar it pops up a modal with more information about the selected activity.
- Create an input to select what program calendar the user wants to see.

## Checklist

- [x] Are the issues being addressed linked to this PR?
- [x] Do all commit messages start with the issue number?
- [x] Are all code changes sufficiently tested?
- [x] Are there screenshots for UI changes?

### CODE REVIEW

**Comment by seidior** [comment]
Is there a concept of an all-day activity within `react-big-calendar`? How are we handling the different activities that are all-day events (like assignments) versus timed events (like class, standup, or retrospective)?

File: webapp/src/components/CalendarPage.tsx
Comment (seidior) [inline webapp/src/components/CalendarPage.tsx:6]: Also, I think that we're going to need to incorporate a <Program> component on the <CalendarPage> even if we're only showing one Program, we should at least see details about the program.

`Program.tsx` and `Activity.tsx` were added from #427, so you'll need to decide to either adapt the code in those files to work with the code you and Prabh wrote for #428 or to delete those files. There's no sense in keeping them around if they're not helpful for what we need to accomplish #428.

File: webapp/src/components/CalendarPage.tsx
Comment (seidior) [inline webapp/src/components/CalendarPage.tsx:10]: Can we clean up some of these imports so they are sorted and separated between package imports and imports of other files within our webapp, similar to how other files have separated out these files?

File: webapp/src/components/CalendarPage.tsx
Comment (seidior) [inline webapp/src/components/CalendarPage.tsx:13]: Perhaps we should set this to "America/Vancouver"?

File: webapp/src/components/CalendarPage.tsx
Comment (seidior) [inline webapp/src/components/CalendarPage.tsx:35]: We should figure out a way that allows us to conform our activities to `react-big-calendar`'s `Event` type so that we can use them directly instead of having to transform them. Or if we can't, some way to make our front-end code work better with the native types within `react-big-calendar`.

**Comment by wendyyng** [comment]
@seidior Optional in #428 to add a drop-down to select which program the user would like to view - is this something that we should expect from this PR? We only have one default program so I think showing a drop-down is not doing much for now but is beneficial if we are handling multiple programs.

File: webapp/src/components/CalendarPage.tsx
Comment (veranika-karpava) [inline webapp/src/components/CalendarPage.tsx:41]: Do we need this line in Calendar? I think we didn't use it.

**Review by veranika-karpava — CHANGES_REQUESTED** [review]
Great job! Everything works as expected in my local machine. 

Also, I think that activities that the duration equal to 0 and end_time: 24:00:00 do not display correctly in week, day and agenda. It doesn't related on this ticket and it'll be the next step
Week:
Screenshot 2022-11-22 at 16 34 23
Day: 
Screenshot 2022-11-22 at 16 35 50
Agenda: 
Screenshot 2022-11-22 at 16 36 46

**Review by phoebehala — COMMENTED** [review]
Some issue about start time and end time might need to further discuss...
For events over than one day such as "Collaborative Work Period", "Independent Work Time", noticed it displayed '>>' in agenda view
image
image

For all-day events such as "Self-assessment", "Prototyping" (start time and end time set to the same), noticed it only displayed 12am .. wondering if the end time need to be changed to 23:59?
image
image

File: webapp/src/components/CalendarPage.tsx
Comment (camianderson) [reply webapp/src/components/CalendarPage.tsx:41]: That is the information that is feeding the <Calendar>

File: webapp/src/components/CalendarPage.tsx
Comment (camianderson) [reply webapp/src/components/CalendarPage.tsx:10]: Great idea!

File: webapp/src/components/CalendarPage.tsx
Comment (camianderson) [reply webapp/src/components/CalendarPage.tsx:13]: I will do it!

**Comment by seidior** [comment]
> @seidior Optional in #428 to add a drop-down to select which program the user would like to view - is this something that we should expect from this PR? We only have one default program so I think showing a drop-down is not doing much for now but is beneficial if we are handling multiple programs.

Not required for this ticket; I wouldn't have objected if it were here but it's not a problem that it isn't. I did mention in a comment on this PR that we should at least be showing some details of the program (like its name and maybe a start/end date), but I think the dropdown might be a good option for individual tickets, which we'll talk about more on Friday/Monday. Good question, @wendyyng!

File: webapp/src/components/CalendarPage.tsx
Comment (veranika-karpava) [reply webapp/src/components/CalendarPage.tsx:41]: Thank you, Cami, the explanation!

**Comment by wendyyng** [comment]
We found an error in the V20 migration. Line 99:  ('Independent Work Time', 'Work on the assigned ticket.', 6, 2, '11:00:00', '24:00:00', 780, 1, @curriculumId), should be week 6 day 4 instead of day 2.

**Comment by wendyyng** [comment]
Further discussion needed for time zone settings
```
Settings.defaultZone = 'America/Vancouver';
const localizer = luxonLocalizer(DateTime);
```

**Comment by wendyyng** [comment]
> Some issue about start time and end time might need to further discuss... For events over than one day such as "Collaborative Work Period", "Independent Work Time", noticed it displayed '>>' in agenda view image image
> 
> For all-day events such as "Self-assessment", "Prototyping" (start time and end time set to the same), noticed it only displayed 12am .. wondering if the end time need to be changed to 23:59? image image

'>>' in agenda view means until the end of day. So I think it's showing tht time correctly. It might be confusing to the users though.

For the all day activities, allDay prop is added to the event. If the duration is 0, we know it's an all day event.
```
const programEventsActivities: RBCEvent[] = activities.map(activity => {
    return {
      title: activity.title,
      start: new Date(activity.start_time),
      end: new Date(activity.end_time),
      description: activity.description_text,
      allDay: activity.duration === 0,
    };
  }
```
Documentation: http://jquense.github.io/react-big-calendar/examples/index.html?path=/docs/props--all-day-accessor

**Comment by wendyyng** [comment]
The time zone settings has been removed for now since it is optional.
`Settings.defaultZone = 'America/Vancouver';`

**Comment by wendyyng** [comment]
> Great job! Everything works as expected in my local machine.
> 
> Also, I think that activities that the duration equal to 0 and end_time: 24:00:00 do not display correctly in week, day and agenda. It doesn't related on this ticket and it'll be the next step Week: Screenshot 2022-11-22 at 16 34 23 Day: Screenshot 2022-11-22 at 16 35 50 Agenda: Screenshot 2022-11-22 at 16 36 46



> activities that the duration equal to 0 and end_time: 24:00:00 do not display correctly in week, day and agenda.

Settings.defaultZone = 'America/Vancouver';

> Great job! Everything works as expected in my local machine.
> 
> Also, I think that activities that the duration equal to 0 and end_time: 24:00:00 do not display correctly in week, day and agenda. It doesn't related on this ticket and it'll be the next step Week: Screenshot 2022-11-22 at 16 34 23 Day: Screenshot 2022-11-22 at 16 35 50 Agenda: Screenshot 2022-11-22 at 16 36 46

@veranika-karpava 
Activities that the duration equal to 0 and end_time: 24:00:00 do not display correctly in week, day and agenda:
- All day event with duration of 0 is showing correctly
- Events with end_time 24:00 are showing correctly in agenda view but not in week & day views. We would need to fix this in the migration instead of this PR - based on my discussion with Spencer

**Comment by seidior** [comment]
program_activities_wednesday

I think I've addressed all the outstanding comments with my last contributed commits. Please pull the latest copy to your local, bring all your Docker containers down, rebuild all Docker containers, and bring them back up. Log in and check out the calendar page.

These latest commits should fix the following:

- Weird display of events that started at a certain time but went until the end of the day.

   These events were being displayed correctly, but they didn't make sense within the context of the program activities. The migration file was updated to address this issue and I force-merged those changes so as not to hold up this PR while I'm away.
   
- If there is more than one program, the program selector and the calendar will be displayed per program, causing multiple calendars to be displayed on the page.

   I've addressed this by moving around the code that was written earlier today by the team into different files so that nothing should get duplicated now. Go ahead, check it out! I actually implemented working functionality for the drop-down selector to update the calendar with different activities for different programs as @wendyyng suggested, and I added a second program in the migration file so that you'd be able to see this functionality work as expected.
   
- Some event durations do not match with (`end_time` - `start_time`)

   This is expected behavior; however, I made sure that my in my updates to the migration file, every duration matches the time between the start time and end time.

- The default time zone should be `America/Vancouver`.

   Previously this was being set in code in the frontend and also in the migration file. Wendy removed the line of code in the webapp, and the migration file was updated to use `America/Vancouver` instead.
   
- All day events were not showing as all day events.

   @wendyyng addressed this particular issue, and I've modified the code style slightly to the same end result.
   
- The week/day view didn't necessarily line up with the agenda view.

   With @wendyyng's fixes for allDay events and with my modifications to our migration file, all displays should appear the same. Of course, leave a comment and let us know if it doesn't.

I've also added some contributions of my own:

- In the process of working on #433, we found out the calendar has a way to scroll down automatically to a certain hour when looking at week or day view. The calendar now shows those views starting at 8:00AM (local time to your computer) instead of 12:00AM, but you can always scroll up inside the calendar to see earlier events if those were ever configured.
- I switched to week view by default, but we can switch it to anything we want. I'm open to good arguments for/against week view by default, and I've made it fairly easy to change this in the code if need be.
- I made the dropdown inline with the title (using MUI's version of a flexbox) to make the entire component look more cohesive. I'd love feedback on the change.

Final notes:

- The changes (and the base code before the changes) don't support mutability of the underlying program or its activities, so this will need to be addressed before adding any future "add activity" or "create program" features.
- There's more work to be done further enhancing the display of the calendar events and programs, but I think what we have now looks good.
- We may want to look into adding "time zone" as a user setting, but that change would need a database migration, changes to `api`, and changes to `webapp`.
- Thank you for the great discussion, @veranika-karpava, @phoebehala, and @wendyyng, and thanks for the leg work on making the calendar happen in the first place, @camianderson and @Prabhj33t. Thanks also to @SanazDn for catching up quickly, contributing, and testing during the day today and providing a fresh perspective. Unfortunately, that means more than half the cohort has worked on this ticket, so the number of people who can officially review this has dwindled in size. But regardless if you've worked on this issue or not, please do take a look at the functionality, the code changes, and the UI changes, and share any thoughts you might have, "official" approval or otherwise.

**Review by veranika-karpava — APPROVED** [review]
Amazing work! Everything works locally.
- checked yarn delint, it's ok
- checked yarn test, all passed
- checked web presentation in browser. Everything works as expected.
Also, I see that short title in week presentation doesn't wrap it to the next line. I think it should the separate issue.
Screenshot 2022-11-24 at 11 00 46

**Review by daveytech — APPROVED** [review]
Great solutions to incorporate the particularities for the package. I believe there are more opportunities to customize the Calendar but can be worked on through maintenance tickets. 

This PR satisfies issue #428 .

**Pull request merged**
